File: languages/cpp/assessments/01-foundations/main.cpp
#include <iostream>
#include <limits>
#include <string>
#include <vector>

using namespace std;

struct Student {
    string name;
    int score;
};

int readPositiveCount() {
    int count = 0;
    while (true) {
        cout << "How many students? ";
        if (!(cin >> count)) {
            cout << "Invalid number. Try again.\n";
            cin.clear();
            cin.ignore(numeric_limits<streamsize>::max(), '\n');
            continue;
        }
        if (count <= 0) {
            cout << "Please enter a positive number.\n";
            cin.ignore(numeric_limits<streamsize>::max(), '\n');
            continue;
        }

        cin.ignore(numeric_limits<streamsize>::max(), '\n');
        return count;
    }
}

int readScore(const string& studentName) {
    int score = 0;
    while (true) {
        cout << "Score for " << studentName << " (0-100): ";
        if (!(cin >> score)) {
            cout << "Invalid score. Enter a number from 0 to 100.\n";
            cin.clear();
            cin.ignore(numeric_limits<streamsize>::max(), '\n');
            continue;
        }
        if (score < 0 || score > 100) {
            cout << "Score out of range. Enter a value from 0 to 100.\n";
            cin.ignore(numeric_limits<streamsize>::max(), '\n');
            continue;
        }

        cin.ignore(numeric_limits<streamsize>::max(), '\n');
        return score;
    }
}

int main() {
    const int count = readPositiveCount();
    vector<Student> students;
    students.reserve(static_cast<size_t>(count));

    for (int i = 0; i < count; ++i) {
        Student student;
        cout << "Student name " << (i + 1) << ": ";
        getline(cin, student.name);
        student.score = readScore(student.name);
        students.push_back(student);
    }

    int total = 0;
    int passCount = 0;
    size_t highestIndex = 0;
    size_t lowestIndex = 0;

    for (size_t i = 0; i < students.size(); ++i) {
        total += students[i].score;
        if (students[i].score >= 60) {
            ++passCount;
        }
        if (students[i].score > students[highestIndex].score) {
            highestIndex = i;
        }
        if (students[i].score < students[lowestIndex].score) {
            lowestIndex = i;
        }
    }

    const double average = static_cast<double>(total) / students.size();

    cout << "\nStudents:\n";
    for (const Student& student : students) {
        cout << "- " << student.name << ": " << student.score << '\n';
    }

    cout << "Average: " << average << '\n';
    cout << "Highest: " << students[highestIndex].name
         << " (" << students[highestIndex].score << ")\n";
    cout << "Lowest: " << students[lowestIndex].name
         << " (" << students[lowestIndex].score << ")\n";
    cout << "Passed: " << passCount << "/" << students.size() << '\n';

    return 0;
}

File: languages/java/assessments/01-foundations/Main.java
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Scanner;

public class Main {
    record Student(String name, int score) {
    }

    static int readPositiveCount(Scanner scanner) {
        while (true) {
            System.out.print("How many students? ");
            String raw = scanner.nextLine().trim();
            try {
                int value = Integer.parseInt(raw);
                if (value > 0) {
                    return value;
                }
                System.out.println("Please enter a number greater than zero.");
            } catch (NumberFormatException error) {
                System.out.println("Please enter a valid integer.");
            }
        }
    }

    static String readNonEmpty(Scanner scanner, String prompt) {
        while (true) {
            System.out.print(prompt);
            String value = scanner.nextLine().trim();
            if (!value.isEmpty()) {
                return value;
            }
            System.out.println("This value cannot be empty.");
        }
    }

    static int readScore(Scanner scanner, String prompt) {
        while (true) {
            System.out.print(prompt);
            String raw = scanner.nextLine().trim();
            try {
                int score = Integer.parseInt(raw);
                if (score >= 0 && score <= 100) {
                    return score;
                }
                System.out.println("Score must be between 0 and 100.");
            } catch (NumberFormatException error) {
                System.out.println("Please enter a valid integer score.");
            }
        }
    }

    public static void main(String[] args) {
        Locale.setDefault(Locale.US);
        Scanner scanner = new Scanner(System.in);
        List<Student> students = new ArrayList<>();
        int count = readPositiveCount(scanner);

        for (int index = 1; index <= count; index++) {
            String name = readNonEmpty(scanner, "Student #" + index + " name: ");
            int score = readScore(scanner, name + "'s score (0-100): ");
            students.add(new Student(name, score));
        }

        int total = 0;
        Student highest = students.get(0);
        Student lowest = students.get(0);
        int passed = 0;

        System.out.println("Students:");
        for (Student student : students) {
            System.out.println("- " + student.name() + ": " + student.score());
            total += student.score();
            if (student.score() > highest.score()) {
                highest = student;
            }
            if (student.score() < lowest.score()) {
                lowest = student;
            }
            if (student.score() >= 60) {
                passed++;
            }
        }

        System.out.printf("Average: %.2f%n", (double) total / students.size());
        System.out.println("Highest: " + highest.name() + " (" + highest.score() + ")");
        System.out.println("Lowest: " + lowest.name() + " (" + lowest.score() + ")");
        System.out.println("Passed: " + passed + "/" + students.size());
    }
}
